composer/trainer/trainer.py

```diff
@@ -1367,9 +1366,14 @@ def _train_loop(self) -> None:
                         full_loss = total_loss.cpu().item()
                         self.logger.data_batch({'loss/train': full_loss / dist.get_world_size()})
 
+                    total_num_samples, total_num_tokens = self._accumulate_samples_and_tokens_across_ranks(
+                        num_samples=self.state.batch_num_samples,
+                        num_tokens=self.state.batch_num_tokens,
+                    )
+
```

Comment:
we're moving this `all_reduce` from `Event.AFTER_DATALOADER` to `Event.BATCH_START`. is this intentional?

Reply:
Good catch; moved back to where it was. However, that's currently a bug -- either it should be before AFTER_DATALOADER or after BATCH_START, depending on whether the timestamp should count the number of samples yielded from the dataloader or the number of samples actually trained. If the latter, then the `batch_num_samples` and `batch_num_token` counts should be updated again after AFTER_DATALOADER and BATCH_START to reflect how algorithms may change the sample count (e.g. selective backprop)